node/src/braid/mod.rs

```
@@ -17,99 +19,144 @@ pub enum AddBeadStatus {
 
-    pub fn add_bead(&mut self, bead: Bead) -> AddBeadStatus {
-        if self.beads.contains(&bead.block_header.block_hash()) {
-            return AddBeadStatus::DagAlreadyContainsBead;
+        for (index, bead) in genesis_beads.into_iter().enumerate() {
+            beads.push(bead.clone());
+            bead_indices.insert(index);
+            bead_index_mapping.insert(bead.block_header.block_hash(), index);
         }
 
-        self.cohorts = self.calculate_cohorts();
-
-        AddBeadStatus::BeadAdded
+        Braid {
+            beads,
+            tips: bead_indices.clone(),
+            cohorts: vec![Cohort(bead_indices.clone())],
+            orphan_beads: Vec::new(),
+            genesis_beads: bead_indices,
+            bead_index_mapping,
+        }
     }
 }
-
+#[allow(unused)]
 impl Braid {
     /// Attempts to extend the braid with the given bead.
     /// Returns true if the bead successfully extended the braid, false otherwise.
-    fn extend(&mut self, bead: &Bead) -> bool {
+    pub fn extend(&mut self, bead: &Bead) -> bool {
         // No parents: bad block
         if bead.committed_metadata.parents.is_empty() {
             return false;
         }
         // Don't have all parents
-        if !bead
-            .committed_metadata
-            .parents
-            .iter()
-            .all(|p| self.beads.contains(p))
-        {
-            return false;
+        for parent_hash in &bead.committed_metadata.parents {
+            // Check if we already have this parent in our beads
+            let parent_exists = self
+                .beads
+                .iter()
+                .any(|b| b.block_header.block_hash() == *parent_hash);
+
+            if !parent_exists {
+                // Try to retrieve the parent
+                if let Some(retrieved_bead) = retrieve_bead(*parent_hash) {
+                    self.extend(&retrieved_bead);
+                } else {
+                    // Parent not found and can't be retrieved
```

This is a silent failure. Here we need to be adding this bead to `orphan_beads`.

At the end of `extend()` we need to be processing anything in `orphan_beads` in case a missing parent was just added.